[feat] Improve frontend audio infrastructure for chat integration

sfc-gh-nbellante opened the pull request:
## Describe your changes

This PR improves the audio recording component by adding proper TypeScript types and enhancing the WaveformController API. Key changes include:

1. Added proper TypeScript interfaces for WaveSurfer and Record plugin mocks in tests
2. Enhanced the WaveformController API with:
   - A new `destroy()` method for proper cleanup
   - A `mountRef` property to allow the component to manage its own container
   - Improved return types for the `stop()` method, now returning metadata along with the blob
3. Fixed memory management by properly cleaning up resources
4. Made the `onApprove` event handler support async operations
5. Improved error handling and initialization logic

## Testing Plan

- Existing tests have been updated to accommodate the new API changes
- Type safety has been improved throughout the codebase
- The changes are well-covered by the existing test suite, which verifies the functionality of the audio recording component

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

sfc-gh-nbellante committed:
Improve frontend audio infrastructure for chat integration

sfc-gh-nbellante commented:
* **#12833** <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12833?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>: 2  dependent PRs ([#12834](https://github.com/streamlit/streamlit/pull/12834) <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12834?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>, [#12835](https://github.com/streamlit/streamlit/pull/12835) <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12835?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>) 👈 <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12833?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#12832** <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12832?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `develop`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

sfc-gh-nbellante committed:
Allow onApprove to return Promise for async operations

- Update WaveformControllerEvents type to allow Promise<void> return
- Await onApprove callback in useWaveformController

sfc-gh-nbellante committed:
Add void operator to promise calls in AudioInput tests

- Mark all onApprove callback invocations with void operator
- Ensures promises are explicitly handled per linting rules

sfc-gh-nbellante commented on frontend/lib/src/components/audio/core/types.ts:
```diff
@@ -14,33 +14,50 @@
   onRecordStart?: () => void
   onRecordReady?: (blob: Blob) => void
-  onApprove?: (wav: Blob) => void
+  onApprove?: (wav: Blob) => void | Promise<void>
```
question: how is this sometimes a promise, and other times not a promise. Can we just make it a promise all the time?

sfc-gh-nbellante commented on frontend/lib/src/components/audio/core/useWaveformController.ts:
```diff
@@ -63,6 +65,9 @@ export function useWaveformController({
   const theme = useEmotionTheme()
 
+  const internalMountRef = useRef<HTMLDivElement>(null)
+  const mountRef = containerRef ?? internalMountRef
```
question: why do we allow an internalMountRef? More specifically, what is different between ChatInput and AudioInput that merits forking this pattern here?

sfc-gh-nbellante commented on frontend/lib/src/components/audio/core/useWaveformController.ts:
```diff
@@ -144,12 +189,10 @@ export function useWaveformController({
+        container: mountRef.current,
         waveColor: theme.colors.primary,
         progressColor: theme.colors.bodyText,
-        height:
```
visual regression suspected: It looks like previously this was adding WAVEFORM_PADDING to the styles. When we switched to `'auto'` I believe this was forgone. How can we ensure this padding gets added in the most idiomatic way?

sfc-gh-nbellante commented on frontend/lib/src/components/audio/core/useWaveformController.ts:
```diff
@@ -189,32 +232,25 @@ export function useWaveformController({
   useEffect(() => {
-    void initializeWaveSurfer()
-    const readyResolvers = readyResolversRef.current
+    initializeWaveSurfer().catch(error => {
```
frustration/concern: why are we doing this pattern where you make initializeWaveSurfer() throw the error and report it to the event handler? If I'm reading this right, the error will get sent to the onError twice, once in initialize itself, another in the .catch(). I'd prefer to not need to .catch() on initializeWaveSurfer unless there's something I'm missing..

sfc-gh-nbellante commented on frontend/lib/src/components/audio/core/useWaveformController.ts:
```diff
@@ -273,6 +309,7 @@ export function useWaveformController({
     }
 
     isPlaybackModeRef.current = false
+    lastStopResultRef.current = null
```
question: what is this `lastStopResultRef` used for?

sfc-gh-nbellante committed:
Make onApprove callback consistently async

Change the onApprove callback signature from `void | Promise<void>`
to always return `Promise<void>` for API consistency.

Changes:
- Update WaveformControllerEvents type signature
- Make AudioInput onApprove implementation async with await
- Update test type annotations

This provides a consistent async interface and prevents bugs from
assuming synchronous behavior when the callback may be async.

sfc-gh-nbellante committed:
Remove internal ref pattern to match ChatInput consistency

Remove the internalMountRef fallback and make containerRef required
in useWaveformController to maintain architectural consistency with
ChatInput.

Changes:
- Remove internalMountRef and fallback logic
- Make containerRef required in UseWaveformControllerParams
- Replace all mountRef references with containerRef
- Return containerRef as mountRef in the controller interface

This maintains API compatibility while enforcing consistent ref
management patterns across the codebase.

sfc-gh-nbellante committed:
Fix visual regression: restore waveform padding

Add vertical padding to StyledWaveSurferDiv to restore spacing that
was previously baked into the height calculation when switching from
explicit height to 'auto'.

Changes:
- Add paddingTop and paddingBottom using theme.spacing.threeXS
- Replaces the previous approach where WAVEFORM_PADDING was included
  in the height calculation (height = WAVEFORM_HEIGHT + WAVEFORM_PADDING)

This properly separates concerns - WaveSurfer handles waveform
dimensions while CSS handles spacing.

sfc-gh-nbellante committed:
Simplify error handling to prevent double error reporting

Remove the throw and .catch() pattern that was causing errors to be
reported twice - once in initializeWaveSurfer's catch block and again
in the useEffect's catch handler.

Changes:
- Remove throw from initializeWaveSurfer catch block
- Remove .catch() from useEffect
- Handle errors once in initializeWaveSurfer where they occur

This creates a single, clear error handling point and prevents
duplicate error reporting to the onError callback.

sfc-gh-nbellante committed:
Make event callbacks consistently return Promise<void>

All WaveformControllerEvents callbacks now consistently return Promise<void>
instead of mixing void and Promise<void> return types. This provides a more
predictable API and aligns with modern async/await best practices where
"always async" is preferred over "sometimes async".

sfc-gh-nbellante committed:
Remove overly defensive lastStopResultRef

The lastStopResultRef was caching stop results as a defensive fallback for:
1. Returning cached results when stop() called repeatedly
2. Providing fallback blob in approve() when state is cleared

This is overly defensive and hides bugs in state management. If stop() is
called when not recording, that's a bug. If approve() can't find the blob,
that's also a bug. Both should fail loudly rather than return stale data.

Simplified approach:
- stop() now throws immediately if not recording
- approve() uses blob parameter or currentBlob only
- Removed all lastStopResultRef cleanup code

This forces proper state management and makes bugs visible.

sfc-gh-nbellante committed:
Replace 'as any' with more specific type assertion in tests

Replaced all `MockRecordPluginClass as any` assertions with
`MockRecordPluginClass as unknown as typeof RecordPlugin` to be more
explicit about the type we're casting to.

While still a type assertion, this is more specific and documents what
type the mock is pretending to be, rather than completely bypassing
type safety with 'as any'.

Also removed the eslint-disable-next-line comments that were suppressing
the no-explicit-any rule, since we're no longer using 'as any'.

sfc-gh-nbellante committed:
Fix type error: import RecordPlugin in test file

Add missing RecordPlugin type import to WaveSurferRecordBackend.test.ts.
The type was referenced in type assertions (typeof RecordPlugin) but
never imported, causing CI type check failures.

sfc-gh-nbellante committed:
Fix TypeScript linting errors in AudioInput components

Resolves 25 ESLint violations that were blocking CI:

**AudioInput.tsx (8 fixes):**
- Add return Promise.resolve() to async callbacks that don't await
- Satisfies @typescript-eslint/require-await rule
- Callbacks: onPermissionDenied, onError, onRecordStart, onRecordReady,
  onCancel, onProgressMs, onPlaybackPause, onPlaybackFinish
- Maintains Promise<void> interface contract while avoiding unnecessary
  async overhead

**AudioInput.test.tsx (17 fixes):**
- Add void operator to unhandled promise calls
- Satisfies @typescript-eslint/no-floating-promises rule
- Makes explicit that promises are intentionally not awaited
- Matches pattern used in useWaveformController implementation

These changes maintain the architectural decision (commit 56221c8d0) to
have all WaveformControllerEvents callbacks return Promise<void> for
API consistency, while satisfying linting requirements.

No behavior changes - callbacks already worked correctly.

sfc-gh-nbellante commented on frontend/lib/src/components/audio/core/useWaveformController.ts:
```diff
@@ -352,14 +376,28 @@ export function useWaveformController({
 
-      eventsRef.current.onRecordReady?.(rawBlob)
-      return rawBlob
+      const durationMs = playerRef.current?.getDuration() ?? 0
```
Good catch on checking this! The code is actually correct here. The WaveSurferPlayer wrapper class already handles the seconds-to-milliseconds conversion in its getDuration() method.

sfc-gh-nbellante committed:
Fix WaveformControllerEvents: make only async callbacks return Promise

Only onApprove is truly async (handles encoding/uploading), so it's the
only callback that should return Promise<void>. All other callbacks are
synchronous (just set React state) and now correctly return void.

This eliminates unnecessary async wrappers and Promise.resolve() calls.

sfc-gh-bnisco commented on frontend/lib/src/components/audio/core/useWaveformController.ts:
```diff
@@ -130,11 +155,26 @@ export function useWaveformController({
+    if (playerRef.current) {
+      configurePlayerEvents(playerRef.current)
+    }
+  }, [events, configurePlayerEvents])
```
**suggestion:** This looks to fall into the case of ["Adjusting Some State When Props Change"](https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes). Therefore, consider using our `useExecuteWhenChanged` hook to handle this instead to remove an unnecessary `useEffect`.

sfc-gh-nbellante replied:
Good catch! I'll refactor this to use `useExecuteWhenChanged` instead of `useEffect`.

sfc-gh-nbellante replied:
Discussed offline, but had to switch back because of side-effect that was breaking. Added more thorough explanation above it in a code comment.

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/AudioInput/styled-components.ts:
```diff
@@ -49,6 +49,9 @@ export const StyledWaveSurferDiv = styled.div<{ show: boolean }>(
     // as sibling divs. Stack them with relative positioning so they overlap.
     position: "relative",
     height: theme.sizes.largestElementHeight,
+    // Add vertical padding to restore spacing that was previously baked into height calculation
```
**suggestion:** Update this comment to make the wording less temporal. "Previously backed into height calculation" won't make sense in the codebase over time. Instead, the comments should reflect exactly what is this accomplishing and why.

sfc-gh-nbellante replied:
Good call, I'll update this to clearly explain what the padding accomplishes and why it's necessary, without referencing the previous implementation.

sfc-gh-nbellante committed:
Address PR feedback: improve React patterns and code clarity

- Replace useEffect with useExecuteWhenChanged in useWaveformController
  to follow React best practices for adjusting state on prop changes
- Update temporal comment wording in AudioInput styled-components
  to improve long-term code maintainability

sfc-gh-nbellante commented:
### Merge activity

* **Nov 1, 7:31 AM UTC**: @sfc-gh-nbellante merged this pull request with [Graphite](https://app.graphite.dev/github/pr/streamlit/streamlit/12833).
